llvm/lib/CodeGen/AsmPrinter/AccelTable.cpp

```diff
@@ -562,36 +595,58 @@ void llvm::emitDWARF5AccelTable(
     default:
       continue;
     }
-    CUIndex[CU.index()] = Count++;
+    CUIndex[CU.index()] = CUCount++;
     assert(CU.index() == CU.value()->getUniqueID());
     const DwarfCompileUnit *MainCU =
         DD.useSplitDwarf() ? CU.value()->getSkeleton() : CU.value().get();
     CompUnits.push_back(MainCU->getLabelBegin());
   }
 
+  for (const auto &TU : TUSymbols) {
```

Comment:
You could use `enumerate` here, rather than `TUCount`
```
for (const auto &[TU, Index] : enumerate(TUSymbols) {
  TUIndex[TU.UniqueID] = Index;
  TypeUnits.push_back(TU.Label);
}
```

But maybe we don't need this at all? Because all the type units are included, unlike above, where some CUs have to be skipped due to the name table kind?

Oh, actually, maybe /that/ code is missing from here. Or is that checking done elsewhere/earlier? (checking that the CU the TU comes from actually requested Default/Apple names?

Reply:
This goes back to my other comment on the counter. Because we can abandon constructing TUs and instead emit that information in CU the UniqueID for TU is not continuous. Because we set it's ID at DwarfTypeUnit creation in DwarfDebug::addDwarfTypeUnitType. 

```
auto OwnedUnit = std::make_unique<DwarfTypeUnit>(
      CU, Asm, this, &InfoHolder, NumTypeUnitsCreated++, getDwoLineTable(CU));
```
      
```
 // Types referencing entries in the address table cannot be placed in type
    // units.
    if (AddrPool.hasBeenUsed()) {
      AccelTypeUnitsDebugNames.clear();
      // Remove all the types built while building this type.
      // This is pessimistic as some of these types might not be dependent on
      // the type that used an address.
      for (const auto &TU : TypeUnitsToAdd)
        TypeSignatures.erase(TU.second);

      // Construct this type in the CU directly.
      // This is inefficient because all the dependent types will be rebuilt
      // from scratch, including building them in type units, discovering that
      // they depend on addresses, throwing them out and rebuilding them.
      setCurrentDWARF5AccelTable(DWARF5AccelTableKind::CU);
      CU.constructTypeDIE(RefDie, cast<DICompositeType>(CTy));
      return;
    }
```

Reply:
I think I left a comment elsewhere that outlined maybe what we could do - could we assign the unique ID late, when we know it'll be used, like around here:

```
 AccelTypeUnitsDebugNames.convertDieToOffset();
```

The IDs could be assigned before the call, the debug_names finalized (I guess not all type units would have names, though, technically, in an LTO situation, so maybe that makes that difficult) then we would've only used IDs for type units that could have accelerated names, not the discarded ones.